src/storage/redis_storage.rs

```diff
@@ -1,34 +1,247 @@
+        let queue: String = format!("{}.{}", &self.prefix, msg.cmd);
+        let value: String = serde_json::to_string(&msg)?;
+
+        conn.rpush(&queue, &value).await?;
+        conn.ltrim(&queue, 0, self.max_command_messages).await?;
+
+        Ok(())
     }
 
-    async fn forward(&self, _msg: Message) -> anyhow::Result<()> {
-        unimplemented!()
+    async fn forward(&self, msg: Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let value: String = serde_json::to_string(&msg)?;
+
+        // add to backlog
+        let key: String = format!("{}.backlog.{}", &self.prefix, &msg.uid);
+        conn.set_ex(&key, &value, self.ttl_in_seconds).await?;
+
+        // push to forward for every destination
+        let queue: String = format!("{}.system.forward", &self.prefix);
+        for dst in msg.dst {
+            let value: String = format!("{}.{}", &msg.uid, &dst.to_string());
+            conn.rpush(&queue, &value).await?
+        }
+
+        Ok(())
     }
 
-    async fn reply(&self, _msg: Message) -> anyhow::Result<()> {
-        unimplemented!()
+    async fn reply(&self, msg: Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let value: String = serde_json::to_string(&msg)?;
+
+        conn.rpush(&msg.ret, &value).await?;
+        Ok(())
     }
 
-    async fn local(&self) -> anyhow::Result<Message> {
-        unimplemented!()
+    async fn local(&self) -> Result<Message> {
+        let mut conn = self.get_connection().await?;
+        let queue: String = format!("{}.system.local", &self.prefix);
+        let ret: Option<Vec<String>> = conn.blpop(&queue, 0).await?;
+
+        let vec = ret.unwrap();
+        if let [_, value] = &vec[..] {
+            let msg: Message = serde_json::from_str(&value)?;
+            Ok(msg)
+        } else {
+            // better to return an error? of what type?
+            unreachable!()
+        }
     }
 
-    async fn queued(&self) -> anyhow::Result<crate::types::QueuedMessage> {
-        Ok(QueuedMessage::Forward(Message::default()))
+    async fn queued(&self) -> Result<Option<QueuedMessage>> {
+        let mut conn = self.get_connection().await?;
+        let forward_queue: String = format!("{}.system.forward", &self.prefix);
+        let reply_queue: String = format!("{}.system.reply", &self.prefix);
+
+        let queues: Vec<String> = vec![forward_queue, reply_queue];
+        let ret: Option<Vec<String>> = conn.blpop(&queues, 0).await?;
+
+        let vec = ret.unwrap();
+        if let [queue, value] = &vec[..] {
+            match queue {
+                forward_queue => {
+                    // forward queue has only <id>.<dst> pair
+                    // a better way to parse these pair?
+                    let mut iter = value.splitn(2, '.');
+                    let id: String = String::from(iter.next().unwrap());
+                    let dst: usize = iter.next().unwrap().parse().unwrap();
```

Comment:
Never use unwrap in production code. this can fail and then kill the entire server. If some one just pushed an invalid string in this queue it will kill rmb and end up losing method. Instead please use proper validation. If a string does not parse probably you take an action (in this case a log, then try the next one)